man/chgpasswd.8
.\"     Title: chgpasswd
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.70.1 <http://docbook.sf.net/>
.\"      Date: 07/30/2006
.\"    Manual: System Management Commands
.\"    Source: System Management Commands
.\"
.TH "CHGPASSWD" "8" "07/30/2006" "System Management Commands" "System Management Commands"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
chgpasswd \- update group passwords in batch mode
.SH "SYNOPSIS"
.HP 10
\fBchgpasswd\fR [\fIoptions\fR]
.SH "DESCRIPTION"
.PP

\fBchgpasswd\fR
reads a list of group name and password pairs from standard input and uses this information to update a set of existing groups. Each line is of the format:
.PP

\fIgroup_name\fR:\fIpassword\fR
.PP
By default the supplied password must be in clear\-text. Default encryption algorithm is DES.
.PP
This command is intended to be used in a large system environment where many accounts are created at a single time.
.SH "OPTIONS"
.PP
The options which apply to the
\fBchgpasswd\fR
command are:
.TP 3n
\fB\-e\fR, \fB\-\-encrypted\fR
Supplied passwords are in encrypted form.
.TP 3n
\fB\-h\fR, \fB\-\-help\fR
Display help message and exit.
.TP 3n
\fB\-m\fR, \fB\-\-md5\fR
Use MD5 encryption instead of DES when the supplied passwords are not encrypted.
.SH "CAVEATS"
.PP
Remember to set permissions or umask to prevent readability of unencrypted files by other users.
.SH "SEE ALSO"
.PP

\fBgpasswd\fR(1),
\fBgroupadd\fR(8).


NEWS
$Id: NEWS,v 1.492 2007/02/01 20:49:25 kloczek Exp $

shadow-4.0.18.1 -> shadow-4.0.18.2					13-10-2007

*** general:
- usermod: fixed handle -a option (by Benno Schulenberg
  <bensberg@justemail.net>),
- useradd: improved auditing support
  (https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=211659),
- groupadd, groupdel, groupmod, useradd, userdel, usermod: flush nscd cashes
  after close /etc/{group,passwd} files,
- su: If compiled without PAM support, enforce the limits from /etc/limits
  when one of the -, -l, or --login options is set, even if called by root.
*** documentation:
- updated translations: ja, nl, tl.
- groupadd.8, groupmod.8, login.1, useradd.8, userdel.8, usermod.8: grammar
  mistakes and other correctstions (by Schulenberg <bensberg@justemail.net>),

shadow-4.0.18 -> shadow-4.0.18.1					03-08-2006

*** general:
- groupmems: fixed compilation when PAM is disabled
  (by Johannes Winkelmann <jw@smts.ch>),
- fixed missing man pages in dist tar ball necessary on build when
  PAM is disabled.

shadow-4.0.17 -> shadow-4.0.18						01-08-2006

*** general:
- su: fixed set enviroment too early when using PAM, so move it to !USE_PAM
  (patch submitted by Mike Frysinger <vapier@gentoo.org>),
- groupadd, groupmod, useradd, usermod: fixed UID/GID overflow (fixed
  http://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=198920)
- passwd, useradd, usermod: fixed inactive/mindays/warndays/maxdays overflow
  (simillar to RH#198920),
- groupmems: rewrited for use PAM and getopt_long() and now it is enabled
  for build and install (patch by George Kraft <gk4@swbell.net>),
- S/Key: removed assign getpass() to libshadow_getpass() on autoconf level
  (patch by Ulrich Mueller <ulm@kph.uni-mainz.de>; http://bugs.gentoo.org/139966),
- usermod: back to previous -a option semantics and clarify -a behavior
  on documentation level (by Greg Schafer <gschafer@zip.com.au>),
- chsh, groupmod: rewrited for use getopt_long().
- updated translations: ca, cs, da, eu, fr, gl, hu, ko, pl, pt, ru, sv, tr, uk, vi.
*** documentation:
- fr and ru man pages are up to date,
- partially translated sv man pages set added
  (by Daniel Nylander <info@danielnylander.se>),
- pl chage(1), chsh(1), groupmod(8): translation finished.

shadow-4.0.16 -> shadow-4.0.17						10-07-2006

*** general:
- userdel, usermod: fixed segfault on remove home directory when it can't
  be removed; for example when it is /dev/null (fixed http://bugs.gentoo.org/139148),
- improved SELinux detection on autoconf level (based on patch by
  Dan Yefimov <dan@D00M.lightwave.net.ru>),
- removed using private implementation getpass() libc function
  (now getpass() is used also when S/KEY support is enabled),
- move nologin do $(sbindir),
- useradd: fixed mail spool file creation (bug cached by Frans Pop
  <elendil@planet.nl>;
  fixed http://bugs.debian.org/374705),
- updated translations: cs, da, de, ko, nb, nl, pt, ro, ru, sk, sv, vi,
- new translations: dz, km, ne.
*** documentation:
- ru man pages up to date,
- lastlog(8): updated pl translation,
- faillog(5): added missing information about fail_locktime element of
  faillog struct (by Thorsten Kukuk <kukuk@suse.de>),
- updated translations: eu, fr, pl.
- reverted using docbook.sourceforge.net in XSL url.

shadow-4.0.15 -> shadow-4.0.16						05-06-2006

*** general:
- userdel: better fix for old CERT VU#312962 (which was fixed in shadow 4.0.8):
  fixed forgoten checking of the return value from fchown() before
  proceeding with the fchmod() (based on Owl patch prepared by
  Rafal Wojtczuk <nergal@owl.openwall.com>),
- userdel: use login.defs::MAIL_DIR instead hardcoded /var/mail in created
  mailbox path (based on Owl fixes submited
  by Solar Designer <solar@openwall.com>),
- by default do not use libshadow_getpass() as getpass() replacemement.
  Use libshadow_getpass() only when S/KEY support is enabled.
  Current glibc getpass() handles correctly longer than 8 characters
  passwords and libshadow_getpass() is used only because libc getpass()
  do not handles password prompting with echo enabled,
- move login.defs::MD5_CRYPT_ENAB to non-PAM part,
- userdel: rewrited for use getopt_log(),
- install default/template configuration files:
-- if shadow is configured with use PAM install /etc/pam.d/* files,
-- if shadow do not uses PAM install /etc/{limits,login.acces} files,
-- install /etc/login.defs and /etc/default/useradd files,
- fixed handle relative symlinks too in lib/commonio.c
  (merge patch from Fedora),
- properly notify nscd to flush its cache
  (https://bugzilla.redhat.com/bugzilla/186803),
- useradd, usermod: fixes for verify return values mkdir() and chown()
  on copy files (merge 482_libmisc_copydir_check_return_values Debian
  patch),
- login, su (non-PAM variant): export MAIL only when MAIL_CHECK_ENAB
  is enabled (Mike Frysinger <vapier@gentoo.org>),
- pgck, grpck: warn when the members of a group differ in /etc/groups
  and /etc/gshadow (fixed http://bugs.debian.org/75181),
- su: fixed exit with a status 0 when the invoked command is terminated
  by a signal which was not catched
  (fixed by Eero Häkkinen <eero17@bigfoot.com>),
- login: cancel login timeout after authentication so that patient people
  timing out on network directory services can log in with local
  accounts (http://bugs.debian.org/107148),
- chgpasswd: fixes for build correctly with --disable-shadowgrp
  (patch by Johannes Winkelmann <jw@tks6.net>).
- updated translations: cs, da, es, eu, fi, fr, gl, hu, id, pt, ru, sk, sv, vi.
- new translations: hu.
*** documentation:
- new cs man pages: groupmems(8), groupmod(8), grpck(8), gshadow(5)
  (by Miroslav Kure <kurem@upcase.inf.upol.cz>),
- regenerate roff man pages using docbook-style-xsl-1.70.1,
- bunch of cleanups in chfn(1), faillog(8), gpasswd(1), groupadd(8),
  groupmems(8), limits(5), login(1), login.defs(5), newgrp(1), passwd(1),
  passwd(5) and su(1) (by Yuri Kozlov <kozlov.y@gmail.com>),
- update pl vipw(8) man page,
- added chgpasswd(8) ru man page,

src/chgpasswd.c
/*
 * Copyright 1990 - 1994, Julianne Frances Haugh
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <config.h>

#ident "$Id: chgpasswd.c,v 1.5 2006/06/20 20:00:04 kloczek Exp $"

#include <fcntl.h>
#include <getopt.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#ifdef USE_PAM
#include "pam_defs.h"
#endif				/* USE_PAM */
#include "defines.h"
#include "nscd.h"
#include "prototypes.h"
#include "groupio.h"
#ifdef	SHADOWGRP
#include "sgroupio.h"
#endif
/*
 * Global variables
 */
static char *Prog;
static int eflg = 0;
static int md5flg = 0;

#ifdef SHADOWGRP
static int is_shadow_grp;
#endif

/* local function prototypes */
static void usage (void);

/*
 * usage - display usage message and exit
 */
static void usage (void)
{
	fprintf (stderr, _("Usage: chgpasswd [options]\n"
			   "\n"
			   "Options:\n"
			   "  -e, --encrypted	supplied passwords are encrypted\n"
			   "  -h, --help		display this help message and exit\n"
			   "  -m, --md5		use MD5 encryption instead DES when the supplied\n"
			   "			passwords are not encrypted\n"
			   "\n"));
	exit (1);
}

int main (int argc, char **argv)
{
	char buf[BUFSIZ];
	char *name;
	char *newpwd;
	char *cp;

#ifdef	SHADOWGRP
	const struct sgrp *sg;
	struct sgrp newsg;
#endif

	const struct group *gr;
	int errors = 0;
	int line = 0;
	int ok;

#ifdef USE_PAM
	pam_handle_t *pamh = NULL;
	struct passwd *pampw;
	int retval;
#endif
#ifndef SHADOWGRP
	struct group newgr;
#endif

	Prog = Basename (argv[0]);

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	{
		int option_index = 0;
		int c;
		static struct option long_options[] = {
			{"encrypted", no_argument, NULL, 'e'},
			{"help", no_argument, NULL, 'h'},
			{"md5", no_argument, NULL, 'm'},
			{NULL, 0, NULL, '\0'}
		};

		while ((c =
			getopt_long (argc, argv, "ehm", long_options,
				     &option_index)) != -1) {
			switch (c) {
			case 'e':
				eflg = 1;
				break;
			case 'h':
				usage ();
				break;
			case 'm':
				md5flg = 1;
				break;
			case 0:
				/* long option */
				break;
			default:
				usage ();
				break;
			}
		}
	}

#ifdef USE_PAM
	retval = PAM_SUCCESS;

	pampw = getpwuid (getuid ());
	if (pampw == NULL) {
		retval = PAM_USER_UNKNOWN;
	}

	if (retval == PAM_SUCCESS) {
		retval = pam_start ("chpasswd", pampw->pw_name, &conv, &pamh);
	}

	if (retval == PAM_SUCCESS) {
		retval = pam_authenticate (pamh, 0);
		if (retval != PAM_SUCCESS) {
			pam_end (pamh, retval);
		}
	}

	if (retval == PAM_SUCCESS) {
		retval = pam_acct_mgmt (pamh, 0);
		if (retval != PAM_SUCCESS) {
			pam_end (pamh, retval);
		}
	}

	if (retval != PAM_SUCCESS) {
		fprintf (stderr, _("%s: PAM authentication failed\n"), Prog);
		exit (1);
	}
#endif				/* USE_PAM */

	/*
	 * Lock the group file and open it for reading. This will bring
	 * all of the entries into memory where they may be updated.
	 */
	if (!gr_lock ()) {
		fprintf (stderr, _("%s: can't lock group file\n"), Prog);
		exit (1);
	}
	if (!gr_open (O_RDWR)) {
		fprintf (stderr, _("%s: can't open group file\n"), Prog);
		gr_unlock ();
		exit (1);
	}
#ifdef SHADOWGRP
	is_shadow_grp = sgr_file_present ();
	if (is_shadow_grp) {
		if (!sgr_lock ()) {
			fprintf (stderr, _("%s: can't lock gshadow file\n"),
				 Prog);
			gr_unlock ();
			exit (1);
		}
		if (!sgr_open (O_RDWR)) {
			fprintf (stderr, _("%s: can't open shadow file\n"),
				 Prog);
			gr_unlock ();
			sgr_unlock ();
			exit (1);
		}
	}
#endif

	/*
	 * Read each line, separating the group name from the password. The
	 * password entry for each group will be looked up in the appropriate
	 * file (gshadow or group) and the password changed.
	 */
	while (fgets (buf, sizeof buf, stdin) != (char *) 0) {
		line++;
		if ((cp = strrchr (buf, '\n'))) {
			*cp = '\0';
		} else {
			fprintf (stderr, _("%s: line %d: line too long\n"),
				 Prog, line);
			errors++;
			continue;
		}

		/*
		 * The groupname is the first field. It is separated from the
		 * password with a ":" character which is replaced with a
		 * NUL to give the new password. The new password will then
		 * be encrypted in the normal fashion with a new salt
		 * generated, unless the '-e' is given, in which case it is
		 * assumed to already be encrypted.
		 */

		name = buf;
		if ((cp = strchr (name, ':'))) {
			*cp++ = '\0';
		} else {
			fprintf (stderr,
				 _("%s: line %d: missing new password\n"),
				 Prog, line);
			errors++;
			continue;
		}
		newpwd = cp;
		if (!eflg) {
			if (md5flg) {
				char salt[12] = "$1$";

				strcat (salt, crypt_make_salt ());
				cp = pw_encrypt (newpwd, salt);
			} else
				cp = pw_encrypt (newpwd, crypt_make_salt ());
		}

		/*
		 * Get the password file entry for this user. The user must
		 * already exist.
		 */
		gr = gr_locate (name);
		if (!gr) {
			fprintf (stderr,
				 _("%s: line %d: unknown group %s\n"), Prog,
				 line, name);
			errors++;
			continue;
		}
#ifdef SHADOWGRP
		sg = sgr_locate (name);
#endif

		/*
		 * The freshly encrypted new password is merged into the
		 * user's password file entry and the last password change
		 * date is set to the current date.
		 */
#ifdef SHADOWGRP
		newsg = *sg;
		newsg.sg_passwd = cp;
#else
		newgr = *gr;
		newgr.gr_passwd = cp;
#endif

		/* 
		 * The updated password file entry is then put back and will
		 * be written to the password file later, after all the
		 * other entries have been updated as well.
		 */
#ifdef SHADOWGRP
		ok = sgr_update (&newsg);
#else
		ok = gr_update (&newgr);
#endif

		if (!ok) {
			fprintf (stderr,
				 _
				 ("%s: line %d: cannot update password entry\n"),
				 Prog, line);
			errors++;
			continue;
		}
	}

	/*
	 * Any detected errors will cause the entire set of changes to be
	 * aborted. Unlocking the password file will cause all of the
	 * changes to be ignored. Otherwise the file is closed, causing the
	 * changes to be written out all at once, and then unlocked
	 * afterwards.
	 */
	if (errors) {
		fprintf (stderr,
			 _("%s: error detected, changes ignored\n"), Prog);
#ifdef SHADOWGRP
		if (is_shadow_grp)
			sgr_unlock ();
#endif
		gr_unlock ();
		exit (1);
	}
#ifdef SHADOWGRP
	if (is_shadow_grp) {
		if (!sgr_close ()) {
			fprintf (stderr,
				 _("%s: error updating shadow file\n"), Prog);
			gr_unlock ();
			exit (1);
		}
		sgr_unlock ();
	}
#endif
	if (!gr_close ()) {
		fprintf (stderr, _("%s: error updating password file\n"), Prog);
		exit (1);
	}

	nscd_flush_cache ("group");

	gr_unlock ();

#ifdef USE_PAM
	if (retval == PAM_SUCCESS)
		pam_end (pamh, PAM_SUCCESS);
#endif				/* USE_PAM */

	return (0);
}
